Vault Config Sourcer Fixes

paladin-devops opened the pull request:
Fixes #2333.

This PR fixes a bug where an application deployed with the Vault config sourcer plugin using a dynamic secrets engine in Vault was inaccessible. `waypoint exec` would hang forever, the address provided by the URL service would be inaccessible, as well as any other way to access the application outside of the URL service.

A subsequent bug which is also fixed in this PR is that dynamic secrets were previously being overwritten as an empty map when the secret was renewed. This is the normal behavior of the Vault lease API, as that API does not return the data of a secret whose lease is renewed. However, now, if a lease is successfully renewed, the Vault config sourcer plugin will still retain the data from when the secret was originally read.

Little bonus in this PR - the CEB binary will now include the git commit in its version, whereas previously only the last major version was logged as the version - this is the Makefile change.

paladin-devops committed:
Upgrade Vault API and SDK deps to latest.

paladin-devops committed:
Update Makefile to add commit to version # of the CEB.

paladin-devops committed:
Start secret renewer for Vault config sourcer in a goroutine.

Also add debug logs to config sourcer, and use NewLifeTimeWatcher from Vault API because NewRenewer is deprecated.

paladin-devops commented:
Quick, weird update on the behavior here! Immediately after the deployment, the secret is indeed set in my environment. Only after ~10 seconds or so do I run into that warning log, and the value is unset.

Run right after the deployment:
```
% waypoint exec env
Connected to deployment v25
USERNAME=vault-generated-username
PASSWORD=vault-generated-password
```
Nothing returned for attempts after ~ 10 seconds post-deployment:
```
% waypoint exec env | grep USERNAME
% waypoint exec env | grep PASSWORD
```

briancain commented on builtin/vault/config_sourcer.go:
```diff
@@ -240,25 +240,31 @@ func (cs *ConfigSourcer) stop() error {
 	// Start the renewer in the background
-	renewer.Renew()
+	go renewer.Renew()
+	log.Debug("renewer started")
```
Can we move these logs to TRACE?

briancain commented on builtin/vault/config_sourcer.go:
```diff
@@ -305,8 +314,9 @@ func (cs *ConfigSourcer) startRenewer(client *vaultapi.Client, path string, s *v
 			}
 
 			cs.cacheMu.Unlock()
+			log.Debug("secret renewed")
```
These all seem good to log at TRACE as well

briancain commented on builtin/vault/config_sourcer.go:
```diff
@@ -240,25 +240,31 @@ func (cs *ConfigSourcer) stop() error {
 
-	renewer, err := client.NewRenewer(&vaultapi.RenewerInput{
+	log.Debug("creating new lifetime watcher...")
+	renewer, err := client.NewLifetimeWatcher(&vaultapi.LifetimeWatcherInput{
```
I think this is the main change right? It'd be nice if the commit message says what this changes and why it was broken before. I'm not entirely sure what this changes versus the `NewRenewer`

edit: Just peeped the Vault client source. Looks like `NewRenewer` was depreciated and called `NewLifetimeWatcher` anyways, so that seems like a good enough thing to include in the commit message

briancain replied:
Should this function also now be renamed `startLifetimeWatcher`?

paladin-devops replied:
done and done!

briancain reviewed (commented):
I think you still have some outstanding TODOs, but I got some review here anyways! ✨

paladin-devops committed:
Fix Vault config sourcer dynamic renewal, and logging updates.

Renewing a Vault secret's lease does not return the data for the secret, it returns null. Before this commit, the data in the secret cache of the config sourcer was thusly being overwritten to an empty map, purging any dynamic secrets which may have been initially set.

paladin-devops committed:
Rename startRenewer func to startLifetimeWatcher.

The name change accurately reflects the Vault API that it is using.

paladin-devops committed:
Reorder params of startLifetimeWatcher.

Logger is preferred to be one of the first parameters.

paladin-devops committed:
Changelog.

paladin-devops committed:
Simplify anonymous func for watching configs for renewal.

briancain reviewed (approved):
🖖🏻
